[WebGPU] Validation layer reports an error when using shared memory textures

mwyrzykowski opened the pull request:
#### ce3d0fe96a044cb21831d23a0ad4099951d2958e
<pre>
[WebGPU] Validation layer reports an error when using shared memory textures
<a href="https://bugs.webkit.org/show_bug.cgi?id=283467">https://bugs.webkit.org/show_bug.cgi?id=283467</a>
<a href="https://rdar.apple.com/140323587">rdar://140323587</a>

Reviewed by Tadeu Zagallo.

Ensure ioSurface backed MTLTextures are allocated with MTLStorageModeManaged
as this is required on some Macs and MTLStorageModeManaged is always supported
on macOS.

Also align Intel graphics allocations with AMD, otherwise we see
asserts with validation enabled on Intel despite everything
seeming to work just fine on all Intel chips we support.

* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::hasUnifiedMemory const):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
Only create IOSurface backed textures with MTLStorageModeManaged, except
when unified memory and where WebGPU is enabled by default.

Canonical link: <a href="https://commits.webkit.org/287372@main">https://commits.webkit.org/287372@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec7171a79d0972bb5403ff8fc14cecadf1babdc3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77765 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30691 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82414 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29039 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65955 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5099 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60916 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/18872 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; layout-tests (exception)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80831 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50888 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66709 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41217 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48245 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27375 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69370 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24577 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83773 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5144 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3681 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69137 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5300 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66705 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68386 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17338 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12407 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10489 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5092 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5096 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8532 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6869 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -510,6 +510,9 @@ static void setOwnerWithIdentity(id<MTLResourceSPI> resource, auto webProcessID)
 {
+#if PLATFORM(MAC) || PLATFORM(MACCATALYST)
+    textureDescriptor.storageMode = MTLStorageModeManaged;
+#endif
```
Is it correct to use `MTLStorageModeManaged` for macCatalyst builds? From reading the relevant validation code I thought you'd need to use `MTLStorageModeShared`, like on iOS.

Also: why override the `storageMode` here instead of setting it correctly at the different call sites of `newTextureWithDescriptor`? Seems kind of confusing if reading the call site, it looks like it's explicitly selecting a storage mode, but it'll be ignored. (And the fact that the `textureDescriptor` object gets modified is additional potential confusion for callers, if they re-use the object, which some do.)

mwyrzykowski replied:
Oh good point about the Mac / macCatalyst, let me see why we had it that way in other places. I thought it would be a driver thing (below the catalyst abstraction) but will double check. 

Also, yes will fix on Friday to set appropriately at the individual call sites.

mwyrzykowski commented on Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
```diff
@@ -194,7 +194,11 @@ static void generateAValidationError(Device& device, NSString* message, bool gen
     MTLTextureDescriptor *textureDescriptor = [MTLTextureDescriptor texture2DDescriptorWithPixelFormat:Texture::pixelFormat(effectiveFormat) width:width height:height mipmapped:NO];
     textureDescriptor.usage = Texture::usage(descriptor.usage, effectiveFormat);
 #if PLATFORM(MAC) || PLATFORM(MACCATALYST)
```
@heycam confirmed macCatalyst should still use managed when the device does not support unified memory, otherwise an iOS app running on an Intel Mac with AMD graphics card will suffer considerable performance penalty syncing memory across the PCI bus

mwyrzykowski committed:
[WebGPU] Validation layer reports an error when using shared memory textures
https://bugs.webkit.org/show_bug.cgi?id=283467
rdar://140323587

Reviewed by Tadeu Zagallo.

Ensure ioSurface backed MTLTextures are allocated with MTLStorageModeManaged
as this is required on some Macs and MTLStorageModeManaged is always supported
on macOS.

Also align Intel graphics allocations with AMD, otherwise we see
asserts with validation enabled on Intel despite everything
seeming to work just fine on all Intel chips we support.

* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::hasUnifiedMemory const):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
Only create IOSurface backed textures with MTLStorageModeManaged, except
when unified memory and where WebGPU is enabled by default.

Canonical link: https://commits.webkit.org/287372@main
